data-prepper-plugins/saas-source-plugins/microsoft-office365-source/src/main/java/org/opensearch/dataprepper/plugins/source/microsoft_office365/Office365CrawlerClient.java

```diff
@@ -198,56 +189,63 @@ private Record<Event> processAuditLog(Map<String, Object> metadata) throws Offic
         }
     }
 
-    private void writeRecordsWithRetry(final List<Record<Event>> records,
-                                       final Buffer<Record<Event>> buffer,
-                                       final AcknowledgementSet acknowledgementSet) {
-        bufferWriteAttemptsCounter.increment();
-        int retryCount = 0;
-        int currentBackoff = 1000; // Start with 1 second
-        final int maxBackoff = 30000; // Max 30 seconds
-        final int maxRetries = 5;
-
-        while (true) {
+    private void writeRecordWithRetry(final Record<Event> record,
+                                      final Buffer<Record<Event>> buffer,
+                                      final AcknowledgementSet acknowledgementSet) {
+        bufferWriteLatencyTimer.record(() -> {
             try {
-                if (configuration.isAcknowledgments()) {
-                    records.forEach(record -> acknowledgementSet.add(record.getData()));
-                    buffer.writeAll(records, (int) Duration.ofSeconds(BUFFER_TIMEOUT_IN_SECONDS).toMillis());
-                    acknowledgementSet.complete();
-                } else {
-                    buffer.writeAll(records, (int) Duration.ofSeconds(BUFFER_TIMEOUT_IN_SECONDS).toMillis());
-                }
+                bufferWriteAttemptsCounter.increment();
+                int retryCount = 0;
+                int currentBackoff = 1000; // Start with 1 second
+                final int maxBackoff = 30000; // Max 30 seconds
+                final int maxRetries = 5;
+
+                while (true) {
+                    try {
+                        if (configuration.isAcknowledgments()) {
+                            acknowledgementSet.add(record.getData());
+                            buffer.write(record, (int) Duration.ofSeconds(BUFFER_TIMEOUT_IN_SECONDS).toMillis());
+                            acknowledgementSet.complete();
```

Comment:
I do not think this is correct. The "acknowledgementSet.complete();" should be done outside of the loop.

Reply:
Or create an acknowledgementset inside the loop